smarts/core/opendrive_road_network.py

```diff
@@ -0,0 +1,1810 @@
+            closest_llp_on_each_route_road,
+            key=lambda l_lp: np.linalg.norm(
+                vec_2d(l_lp.lp.pose.position) - vec_2d(point)
+            ),
+        )
+        closest_lane = closest_linked_lp.lp.lane
+        waypoint_paths = []
+        for lane in closest_lane.road.lanes:
+            if closest_lane.road.road_id != route[-1] and all(
+                out_lane.road.road_id not in route for out_lane in lane.outgoing_lanes
+            ):
+                continue
+            waypoint_paths += lane._waypoint_paths_at(point, lookahead, route)
+
+        return sorted(waypoint_paths, key=len, reverse=True)
+
+    def _equally_spaced_path(
+        self,
+        path: Sequence[LinkedLanePoint],
+        point: Tuple[float, float, float],
+        lp_spacing: float,
+        width_threshold=None,
+    ) -> List[Waypoint]:
+        """given a list of LanePoints starting near point, return corresponding
+        Waypoints that may not be evenly spaced (due to lane change) but start at point."""
+
+        continuous_variables = [
+            "positions_x",
+            "positions_y",
+            "headings",
+            "lane_width",
+            "speed_limit",
+        ]
+        discrete_variables = ["lane_id", "lane_index"]
+
+        ref_lanepoints_coordinates = {
+            parameter: [] for parameter in (continuous_variables + discrete_variables)
+        }
+        curr_lane_id = None
+        skip_lanepoints = False
+        index_skipped = []
+        for idx, lanepoint in enumerate(path):
+
+            if lanepoint.is_inferred and 0 < idx < len(path) - 1:
+                continue
+
+            if curr_lane_id is None:
+                curr_lane_id = lanepoint.lp.lane.lane_id
+
+            # Compute the lane offset for the lanepoint position
+            position = Point(
+                x=lanepoint.lp.pose.position[0], y=lanepoint.lp.pose.position[1], z=0.0
+            )
+            lane_coord = lanepoint.lp.lane.to_lane_coord(position)
+            # Skip one-third of lanepoints for next lanes not outgoing to previous lane
+            if skip_lanepoints:
+                if lane_coord.s > (lanepoint.lp.lane.length / 3):
+                    skip_lanepoints = False
+                else:
+                    index_skipped.append(idx)
+                    continue
+
+            if lanepoint.lp.lane.lane_id != curr_lane_id:
+                previous_lane = self._lanes[curr_lane_id]
+                curr_lane_id = lanepoint.lp.lane.lane_id
+                # if the current lane is not outgoing to previous lane, start skipping one third of its lanepoints
+                if lanepoint.lp.lane not in previous_lane.outgoing_lanes:
+                    skip_lanepoints = True
+                    index_skipped.append(idx)
+                    continue
+
+            # Compute the lane's width at lanepoint's position
+            width_at_offset = lanepoint.lp.lane.width_at_offset(lane_coord.s)
```

Comment:
This may require some optimization (as it happens for every Waypoint on every path on every step).  We may want to consider moving this back into the `Lanepoint` class?

Reply:
I will create lane_width, attribute of `LanePoint` then.